syncer/handle_error.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pingcap/dm/dm/command"
+	"github.com/pingcap/dm/dm/pb"
+	parserpkg "github.com/pingcap/dm/pkg/parser"
+	"github.com/pingcap/dm/pkg/terror"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	tmysql "github.com/pingcap/parser/mysql"
+
+	"github.com/siddontang/go-mysql/replication"
+)
+
+// HandleError handle error for syncer
+func (s *Syncer) HandleError(ctx context.Context, req *pb.HandleWorkerErrorRequest) (string, error) {
+	pos := req.BinlogPos
+
+	if len(pos) == 0 {
+		startLocation := s.getErrLocation()
+		if startLocation == nil {
+			return fmt.Sprintf("source '%s' has no error", s.cfg.SourceID), nil
+		}
+		pos = startLocation.Position.String()
+	} else {
+		startLocation, err := command.VerifyBinlogPos(pos)
+		if err != nil {
+			return "", err
+		}
+		pos = startLocation.String()
+	}
+
+	events := make([]*replication.BinlogEvent, 0)
+	var err error
+	if req.Op == pb.ErrorOp_Replace {
+		events, err = s.genEvents(req.Sqls)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	err = s.errOperatorHolder.Set(pos, req.Op, events)
+	if err != nil {
+		return "", err
+	}
+
+	// remove outdated operators when add operator
+	err = s.errOperatorHolder.RemoveOutdated(s.checkpoint.FlushedGlobalPoint())
+	if err != nil {
+		return "", err
+	}
+
+	return "", nil
+}
+
+func (s *Syncer) genEvents(sqls []string) ([]*replication.BinlogEvent, error) {
+	events := make([]*replication.BinlogEvent, 0)
+
+	parser2 := parser.New()
+	if s.cfg.EnableANSIQuotes {
+		parser2.SetSQLMode(tmysql.ModeANSIQuotes)
```

Comment:
seems `parser2` is used in many place, how about let it become a menber of `Syncer`

Reply:
Most of them use `getParser(db,enableAnisquotes)` for special db. I think that's better to avoid misuse of `enable_ansiquotes`.